// SNAKE/map.js

module.exports = Map;


function Map(sizeX, sizeY){
	this.maxX = sizeX;
	this.maxY = sizeY;
	this.size = this.maxX * this.maxY;
	this.arr = new Array(this.maxX * this.maxY);
	this.food = [];
	this.initialise();
}

Map.prototype.getCell = function(x, y){
	return this.arr[x + y * this.maxX];
};

Map.prototype.setCell = function(x, y, value){
	this.arr[x + y * this.maxX] = value;
	if(value === "food"){
		this.food.push({x:x,y:y});
	}
	//console.log("INITIAL FOOD"+this.food[0].x+" "+this.food[0].y);
};

Map.prototype.getLength = function(){
	return this.size;
};
Map.prototype.foodEaten = function(position){
	for(var i = 0 ,max = this.food.length; i < max ; i++){
		if(this.food[i].x === position.x && this.food[i].y === position.y){
			this.setCell(this.food[i].x,this.food[i].y,"empty");
			this.food.splice(i,1);

		}
	}
}
Map.prototype.setFood = function(food){
	this.food = food;
}
Map.prototype.initialise = function(){
	for(var i = 0 , max = this.arr.length; i<max; i++){
		this.arr[i] = "empty";
	}
};
Map.prototype.getMapGrid = function(){
	return this.arr;
}
Map.prototype.getFood = function(){
	return this.food;
}
Map.prototype.clearIt = function(){
	for(var i = 0 , max = this.size ; i < max ; i++){
		this.arr[i] = 0;
	}
};
