// SNAKE/player.js
module.exports = Player;
function Player(x,y,size,direction,id) {
	this.snakeArray = [];
	this.initialX = x;
	this.initialY = y;
	this.direction = direction;
	this.growing = 0;
	this.id = id;
	this.initialise();
}
Player.prototype.getId = function(){
	return this.id;
}
Player.prototype.initialise = function(){
	if(this.direction === "LEFT"){
	for(var i = 0 ; i <5 ; i ++){
		//Push adds at the end of the array
		this.snakeArray.push({x:this.initialX+i,y:this.initialY});
		}
	}else if(this.direction ==="RIGHT"){
		for(var i = 0 ; i < 5 ; i++ ){
		//Push adds at the end of the array
		this.snakeArray.push({x:this.initialX-i,y:this.initialY});
		}  
	}
};
Player.prototype.getDirection = function(){
	return this.direction;
}
Player.prototype.setDirection = function(direction){
	if((direction ==="UP" && this.direction!=="DOWN") || 
		(direction === "DOWN" && this.direction!=="UP") ||
		(direction ==="LEFT" && this.direction!=="RIGHT") ||
		(direction ==="RIGHT" && this.direction!=="LEFT")){
		this.direction = direction;
	}else{
		//console.log("Invalid direction :"+direction);
	}
};
Player.prototype.move = function(){
	var headX = this.snakeArray[0].x;
	var headY = this.snakeArray[0].y;
	if(this.direction === "RIGHT"){
		++headX;
	}else if(this.direction === "LEFT"){
		--headX;
	}else if(this.direction === "UP"){
		--headY;
	}else if(this.direction === "DOWN"){
		++headY;
	}
	var newHead = {x:headX,y:headY};
	//Add new head to the array
	this.snakeArray.unshift(newHead);
	//pop the last cell of the snake
	//if it isn't growing
	if(this.growing===0){
	this.snakeArray.pop();
	}else{
		this.growing = this.growing-1;
	}
};
Player.prototype.getGrowth = function(){
	return this.growing;
}
Player.prototype.grow = function(growth){
	this.growing = this.growing+ growth;
};

Player.prototype.getHead = function(){
	return this.snakeArray[0];
};
Player.prototype.getTail = function(){
	return this.snakeArray[this.snakeArray.length-1];
}
Player.prototype.getSnakeArray = function(){
	return this.snakeArray;
}
Player.prototype.setHead = function(position){
	this.snakeArray[0].x = position.x;
	this.snakeArray[0].y = position.y;
	
}
