// SNAKE/waitingRoom.js
var aController = require('SNAKE/controller.js');
module.exports = WaitingRoom;
function WaitingRoom(io){

	this.numberOfUsers = 0;
	this.hello = "This is sparta "+this.numberOfUsers;
	this.io = io;
	this.users =  [];
	this.controllers = [];
	this.playingTogether = [];
}
WaitingRoom.prototype.returnHello = function(){
	return this.hello;
}
WaitingRoom.prototype.oneMoreUser = function(name, socket){
		this.setupSockets(socket);
	++this.numberOfUsers;
	this.users.push({socket:socket,name:name});
	
	if(this.users.length >1){
		this.controllers.push(new aController(60, this.users[0].socket, this.users[1].socket))

		this.playingTogether.push({
			p1: this.users[0].socket.id,
			p2: this.users[1].socket.id
		});
				this.users.splice(0,2);
	}

	

}
WaitingRoom.prototype.setupSockets = function(socket){

}
WaitingRoom.prototype.userLost = function(socket){
		console.log("user lost");
	for(var i = 0 ; i < this.controllers.length; i++){
		var p = this.controllers[i].getPlayers();
		for(var j = 0, max = p.length; j < max ; j++)
			if(p[j] === socket.id){
				this.controllers.splice(i,1);
				console.log("DELETED");
			}
	}

}
